File: python/voxelboom/noise.py
import numpy as np

_U64 = np.uint64


def _splitmix64(x):
    with np.errstate(over="ignore"):
        x = x + _U64(0x9E3779B97F4A7C15)
        z = x
        z = (z ^ (z >> _U64(30))) * _U64(0xBF58476D1CE4E5B9)
        z = (z ^ (z >> _U64(27))) * _U64(0x94D049BB133111EB)
        z = z ^ (z >> _U64(31))
    return z


def hash01(ix, iy, iz, seed):
    with np.errstate(over="ignore"):
        ix = np.asarray(ix, dtype=np.int64).astype(_U64)
        iy = np.asarray(iy, dtype=np.int64).astype(_U64)
        iz = np.asarray(iz, dtype=np.int64).astype(_U64)
        s = _U64(int(seed) & 0xFFFFFFFFFFFFFFFF)
        h = _splitmix64(ix * _U64(0x9E3779B97F4A7C15) ^ s)
        h = _splitmix64(h ^ (iy * _U64(0xC2B2AE3D27D4EB4F)))
        h = _splitmix64(h ^ (iz * _U64(0x165667B19E3779F9)))
    # top 53 bits -> double in [0, 1)
    return (h >> _U64(11)).astype(np.float64) * (1.0 / float(1 << 53))


class ValueNoise3D:

    def __init__(self, seed):
        self.seed = int(seed) & 0x7FFFFFFF

    def _lattice(self, ix, iy, iz):
        return hash01(ix, iy, iz, self.seed) * 2.0 - 1.0

    def sample(self, p):
        p = np.asarray(p, dtype=np.float64)
        x, y, z = p[..., 0], p[..., 1], p[..., 2]
        x0 = np.floor(x)
        y0 = np.floor(y)
        z0 = np.floor(z)
        xf, yf, zf = x - x0, y - y0, z - z0
        x0i, y0i, z0i = x0.astype(np.int64), y0.astype(np.int64), z0.astype(np.int64)
        x1i, y1i, z1i = x0i + 1, y0i + 1, z0i + 1
        # smoothstep fade
        u = xf * xf * (3.0 - 2.0 * xf)
        v = yf * yf * (3.0 - 2.0 * yf)
        w = zf * zf * (3.0 - 2.0 * zf)
        c000 = self._lattice(x0i, y0i, z0i)
        c100 = self._lattice(x1i, y0i, z0i)
        c010 = self._lattice(x0i, y1i, z0i)
        c110 = self._lattice(x1i, y1i, z0i)
        c001 = self._lattice(x0i, y0i, z1i)
        c101 = self._lattice(x1i, y0i, z1i)
        c011 = self._lattice(x0i, y1i, z1i)
        c111 = self._lattice(x1i, y1i, z1i)
        x00 = c000 + u * (c100 - c000)
        x10 = c010 + u * (c110 - c010)
        x01 = c001 + u * (c101 - c001)
        x11 = c011 + u * (c111 - c011)
        y0v = x00 + v * (x10 - x00)
        y1v = x01 + v * (x11 - x01)
        return y0v + w * (y1v - y0v)


def fbm(noise, p, octaves=4, lacunarity=2.0, gain=0.5):
    p = np.asarray(p, dtype=np.float64)
    total = np.zeros(p.shape[:-1], dtype=np.float64)
    amp = 1.0
    freq = 1.0
    norm = 0.0
    for _ in range(octaves):
        total = total + amp * noise.sample(p * freq)
        norm += amp
        amp *= gain
        freq *= lacunarity
    return total / norm

File: java/common/src/main/java/com/example/explosionlib/engine/Noise.java
package com.example.explosionlib.engine;

public final class Noise {
    private Noise() {}

    static long splitmix64(long x) {
        x += 0x9E3779B97F4A7C15L;
        long z = x;
        z = (z ^ (z >>> 30)) * 0xBF58476D1CE4E5B9L;
        z = (z ^ (z >>> 27)) * 0x94D049BB133111EBL;
        return z ^ (z >>> 31);
    }

    public static double hash01(long ix, long iy, long iz, long seed) {
        long h = splitmix64(ix * 0x9E3779B97F4A7C15L ^ seed);
        h = splitmix64(h ^ (iy * 0xC2B2AE3D27D4EB4FL));
        h = splitmix64(h ^ (iz * 0x165667B19E3779F9L));
        return (h >>> 11) * (1.0 / (1L << 53));
    }

    public static final class ValueNoise3D {
        private final long seed;

        public ValueNoise3D(long seed) {
            this.seed = seed & 0x7FFFFFFFL;
        }

        private double lattice(long ix, long iy, long iz) {
            return hash01(ix, iy, iz, seed) * 2.0 - 1.0;
        }

        public double sample(double x, double y, double z) {
            long x0 = (long) Math.floor(x), y0 = (long) Math.floor(y), z0 = (long) Math.floor(z);
            double xf = x - x0, yf = y - y0, zf = z - z0;
            long x1 = x0 + 1, y1 = y0 + 1, z1 = z0 + 1;
            double u = xf * xf * (3.0 - 2.0 * xf);
            double v = yf * yf * (3.0 - 2.0 * yf);
            double w = zf * zf * (3.0 - 2.0 * zf);
            double c000 = lattice(x0, y0, z0), c100 = lattice(x1, y0, z0);
            double c010 = lattice(x0, y1, z0), c110 = lattice(x1, y1, z0);
            double c001 = lattice(x0, y0, z1), c101 = lattice(x1, y0, z1);
            double c011 = lattice(x0, y1, z1), c111 = lattice(x1, y1, z1);
            double x00 = c000 + u * (c100 - c000), x10 = c010 + u * (c110 - c010);
            double x01 = c001 + u * (c101 - c001), x11 = c011 + u * (c111 - c011);
            double y0v = x00 + v * (x10 - x00), y1v = x01 + v * (x11 - x01);
            return y0v + w * (y1v - y0v);
        }

        public double fbm(double x, double y, double z, int octaves) {
            double total = 0.0, amp = 1.0, freq = 1.0, norm = 0.0;
            for (int i = 0; i < octaves; i++) {
                total += amp * sample(x * freq, y * freq, z * freq);
                norm += amp;
                amp *= 0.5;
                freq *= 2.0;
            }
            return total / norm;
        }
    }

    public static final class DetRandom {
        private long s;

        public DetRandom(long seed) {
            this.s = seed;
        }

        public long nextLong() {
            s += 0x9E3779B97F4A7C15L;
            long z = s;
            z = (z ^ (z >>> 30)) * 0xBF58476D1CE4E5B9L;
            z = (z ^ (z >>> 27)) * 0x94D049BB133111EBL;
            return z ^ (z >>> 31);
        }

        public double nextDouble() {
            return (nextLong() >>> 11) * (1.0 / (1L << 53));
        }

        public int nextInt(int bound) {
            return (int) (nextDouble() * bound);
        }

        public double nextGaussian() {
            double u1 = Math.max(nextDouble(), 1e-12);
            double u2 = nextDouble();
            return Math.sqrt(-2.0 * Math.log(u1)) * Math.cos(2.0 * Math.PI * u2);
        }
    }
}
